pkg/local_object_storage/engine/list.go

```diff
@@ -114,7 +114,16 @@ func (e *StorageEngine) ListWithCursor(prm *ListWithCursorPrm) (*ListWithCursorR
 			continue
 		}
 
-		result = append(result, res.AddressList()...)
+		addressList := res.AddressList()
+		for j := range addressList {
+			_, err = e.exists(addressList[j])
+			if err != nil {
+				continue
+			}
```

Comment:
I would do it in parallel. All objects here are _likely_ to exist, so we will process all 4 shards for each object.
Alternatively, we can (1) process all 4 shards with `exists(list of objects)` and (2) even do it in parallel.
In case we are not doing it now, could you add a TODO here?